File: third-party/thrift/src/thrift/conformance/java/src/main/java/com/meta/thrift/conformance/server/ConformanceServer.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.meta.thrift.conformance.server;

import com.facebook.swift.service.ThriftServerConfig;
import com.facebook.thrift.server.RpcServerHandler;
import com.facebook.thrift.server.ServerTransport;
import com.facebook.thrift.util.RpcServerUtils;
import com.facebook.thrift.util.TransportType;
import java.net.InetSocketAddress;
import org.apache.thrift.conformance.ConformanceService;

public class ConformanceServer {

  public static void main(String[] args) {
    RpcServerHandler handler =
        ConformanceService.serverHandlerBuilder(new ConformanceServiceImpl()).build();

    ServerTransport transport =
        RpcServerUtils.createServerTransport(
                new ThriftServerConfig().setPort(0), TransportType.THEADER, handler)
            .block();

    // Conformance test runner starts multiple instances of server
    // and reads the port from std output
    // Ports should be allocated automatically
    System.out.println(((InetSocketAddress) transport.getAddress()).getPort());

    transport.onClose().block();
  }
}

File: third-party/thrift/src/thrift/conformance/cpp2/ConformanceServer.cpp
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <iostream>
#include <memory>

#include <glog/logging.h>
#include <folly/init/Init.h>
#include <thrift/conformance/cpp2/AnyRegistry.h>
#include <thrift/conformance/cpp2/ConformanceHandler.h>
#include <thrift/lib/cpp2/server/ThriftServer.h>

int main(int argc, char** argv) {
  const folly::Init init(&argc, &argv);
  auto server = std::make_shared<apache::thrift::ThriftServer>();
  auto handler =
      std::make_shared<apache::thrift::conformance::ConformanceHandler>();
  server->setInterface(handler);
  server->setPort(0); // Any free port.
  server->setup();
  LOG(INFO)
      << apache::thrift::conformance::AnyRegistry::generated().debugString();
  std::cout << server->getAddress().getPort() << std::endl;
  server->waitForStop();
  return 0;
}
